File: BFGS/Cpp/examples/rosenbrockFunction.cpp
#include <cstring>
#include "struct.h"

// global minimum  : 0
// global minimizer: (1, 1, ..., 1)

void objgrd(double *fcnVal, double *grdVec, double *posVec, fcndata &fcnObj)
{
	int varNum = fcnObj.varNum;

	*fcnVal = 0.0;
	memset(grdVec, 0, sizeof(double) * varNum);

	for ( int varIdx = 0; varIdx < varNum - 1; ++varIdx )
	{
		double x0 = posVec[varIdx    ];
		double x1 = posVec[varIdx + 1];
		double f0 = x1 - x0 * x0;
		double f1 = 1.0 - x0;
		
		*fcnVal += 100.0 * f0 * f0 + f1 * f1;

		grdVec[varIdx    ] += -400.0 * f0 * x0 - 2.0 * f1;
		grdVec[varIdx + 1] +=  200.0 * f0;
	}

	return;
}

File: LBFGS/MATLAB/examples/rosenbrockFunction.m
function [fcnVal, grdVec] = rosenbrockFunction(posVec)

% minimum  : 0
% minimizer: (1, 1, ..., 1)

posNxtVec = posVec(2 :  end     );
posNowVec = posVec(1 : (end - 1));

fcnVal = sum(100 * (posNxtVec - posNowVec.^2).^2 + (1 - posNowVec).^2);

grdVec = zeros(size(posVec));
grdVec(1 : (end - 1)) = 200 * (posNxtVec - posNowVec.^2) .* (-2 * posNowVec) - 2 * (1 - posNowVec);
grdVec(2 :  end     ) = grdVec(2 : end) + 200 * (posNxtVec - posNowVec.^2);
